### PR TITLE
Support sparse adjacency matrix for the saliency map of GCN models. 

### PR DESCRIPTION
**Pull request by sktzwhj**
Hi @adocherty , 

Please have a look at this PR which added the sparse adjacency matrix support in the saliency maps of GCN models. Thanks.

-Huijun

### CODE REVIEW

File: stellargraph/utils/saliency_maps/saliency.py [hunk 1]
```diff
@@ -114,11 +121,14 @@ def get_link_masks(self, X_val, A_val, node_idx, class_of_interest):
...
         if self.is_sparse:
-            raise NotImplementedError("Sparse matrix support is not yet implemented")
-
-        gradients = self.compute_link_gradients(
-            [X_val, out_indices, A_val, 0, class_of_interest]
-        )
+            # raise NotImplementedError("Sparse matrix support is not yet implemented")
+            gradients = self.compute_link_gradients(
+                [X_val, out_indices, A_index, A_val, 0, class_of_interest]
+            )
```
[the remark this answers is not in the document]
Comment (sktzwhj) [reply stellargraph/utils/saliency_maps/saliency.py]: Good suggestion! Fixed.

File: stellargraph/utils/saliency_maps/integrated_gradients.py [hunk 2]
```diff
@@ -100,12 +108,16 @@ def get_integrated_link_masks(
...
-            tmp = self.get_link_masks(X_val, A_step, node_idx, class_of_interest)
+            tmp = self.get_link_masks(
+                X_val, A_index, A_step, node_idx, class_of_interest
+            )
             total_gradients += tmp
 
         return np.squeeze(np.multiply(total_gradients, A_diff) / steps, 0)
 
-    def get_node_importance(self, X_val, A_val, node_idx, class_of_interest, steps=20):
+    def get_node_importance(
```
[the remark this answers is not in the document]
Comment (sktzwhj) [reply stellargraph/utils/saliency_maps/integrated_gradients.py]: Updated.

**Comment by adocherty** [comment]
Mostly the notebook works and the integrated gradient saliency matches the brute force calculated saliency.  However, sometimes the two are quite different, particularly the sign of the edge importances, for example:

Integrated Gradients:
image

Brute force calculation:
image

Increasing the number of steps in IG calculation doesn't change the result. I'm a bit worried by this.  Do you have any feeling for why this is occurring?

**Comment by adocherty** [comment]
It would also be good to add some simple unit tests that check if the arguments are passes incorrectly that the correct assertions are raised. See `test_node_mappers.py` or `test_ensemble.py` for examples.

**Comment by sktzwhj** [comment]
> Mostly the notebook works and the integrated gradient saliency matches the brute force calculated saliency. However, sometimes the two are quite different, particularly the sign of the edge importances, for example:
> 
> Integrated Gradients:
> image
> 
> Brute force calculation:
> image
> 
> Increasing the number of steps in IG calculation doesn't change the result. I'm a bit worried by this. Do you have any feeling for why this is occurring?

I think it's just the nature of 'approximations'. After all, we are using the average of several continuous gradients to approximate the discrete 'gradients'. It may have that the values we sampled along the path do not cover some extreme values. As the match happens in majority cases, I do not think we should worry much. Intuitively, the saliency map for an ensemble of models could resolve this.

**Comment by sktzwhj** [comment]
> I've modified the classes to not require a `sparse` argument and rather to get this from the generator that you pass. Could you check that this is OK?

Thanks, it works fine.

**Comment by sktzwhj** [comment]
> It would also be good to add some simple unit tests that check if the arguments are passes incorrectly that the correct assertions are raised. See `test_node_mappers.py` or `test_ensemble.py` for examples.

I have added a test to do this.

**Comment by adocherty** [comment]
> I think it's just the nature of 'approximations'. After all, we are using the average of several continuous gradients to approximate the discrete 'gradients'. It may happen that the values we sampled along the path do not cover some extreme values. As the match happens in majority cases, I do not think we should worry much. Intuitively, the saliency map for an ensemble of models could resolve this.

If this was the nature of the approximation of the integral by a discrete sum we should see that the values would become closer when we increase the number of steps; but this doesn't happen.  I'm not too worried by the value being different or a small positive value becoming small negative, but here a large (relatively) positive value becomes negative. This I'm troubled by.

I don't think we should investigate now, but we shouldn't forget about this.

**Comment by adocherty** [comment]
Actually, I just noticed you committed this file `demos/interpretability/gcn/hateful-twitters-interpretability.ipynb` which looks like it should be in a separate PR.  Did you mean to do this?

**Comment by sktzwhj** [comment]
Remove demo to create another PR. This PR is still merged as nothing else has been changed.

**Pull request merged**